### PR TITLE
Make side panels resizable

### PR DESCRIPTION
**Pull request by laymonage**
This PR adds a resize grip to the side panel that resizes the side panel when dragged. Works with mouse, touch, and keyboard.

For keyboard accessibility, the side panel is resizable using an invisible `<input type="range" />` that's wrapped inside the "grip". To use it, focus on the grip and press left and right buttons on your keyboard.

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**: Chrome 107, Firefox 106, Safari 15.4
    -   [x] **Please list which assistive technologies [^3] you tested**: VoiceOver on macOS Ventura 13.0.1

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: client/scss/components/_form-side.scss [hunk 1]
```diff
@@ -20,11 +28,11 @@
       w-py-4
       w-bg-white
       w-box-border
-      w-transition
+      w-transition-all
+      motion-reduce:w-transition-none
       w-duration-300
       w-border-l
       w-border-grey-100
-      w-overflow-y-auto
```
Comment (thibaudcolas) [inline client/scss/components/_form-side.scss:27]: I think this will make it impossible to scroll the panels if they’re taller than the viewport. Not sure why this change was needed?

File: client/scss/components/_form-side.scss [hunk 2]
```diff
@@ -49,7 +61,33 @@
...
+
+    .icon {
+      @apply w-w-4 w-h-4;
+    }
+  }
+
+  &__width-input {
+    @apply w-w-0 w-h-0 w-opacity-0 w-absolute w-pointer-events-none;
+
+    direction: rtl;
```
Comment (thibaudcolas) [inline client/scss/components/_form-side.scss:86]: This concerns me a bit since this property is primarily meant to display content written in right-to-left languages – not sure I understand the potential implications of setting this to a form field (for example – would it change how the label is announced to screen reader users?).

Is there another approach we could use which doesn’t rely on this?

File: client/scss/components/_preview-panel.scss [hunk 3]
```diff
@@ -19,12 +19,29 @@
...
   &__wrapper {
+    position: relative;
     width: calc(var(--preview-iframe-width) * var(--preview-width-ratio));
     height: 100%;
     margin-inline-start: auto;
     margin-inline-end: auto;
   }
 
+  // Transparent overlay to prevent drag-resize issues
+  // when the cursor is above the iframe.
```
Comment (thibaudcolas) [inline client/scss/components/_preview-panel.scss:30]: Just checking if you tried something as simple as setting the iframe to `pointer-events: none` as well?

File: client/src/includes/sidePanel.js [hunk 4]
```diff
@@ -99,6 +102,123 @@ export default function initSidePanel() {
...
+      : currentWidth;
+
+    sidePanelWrapper.style.width = `${newWidth}px`;
+    widthInput.value = ((newWidth - minWidth) / range) * 100;
+    localStorage.setItem('wagtail:side-panel-width', newWidth);
+  };
+
+  resizeGrip.addEventListener('dragstart', (e) => {
+    // When dragging, replace 'ghost' image of the element with a blank canvas.
+    // It needs to be in the DOM, so add it temporarily and remove on dragend.
```
Comment (thibaudcolas) [inline client/src/includes/sidePanel.js:132]: Could you explain why? This just explains what the code does but not why we do it.

File: client/src/includes/sidePanel.js [hunk 5]
```diff
@@ -99,6 +102,123 @@ export default function initSidePanel() {
...
+    e.dataTransfer.effectAllowed = 'move';
+    e.dataTransfer.setDragImage(blankCanvas, 0, 0);
+
+    document.body.classList.add('side-panel-resizing');
+  });
+
+  resizeGrip.addEventListener('dragend', () => {
+    document.body.classList.remove('side-panel-resizing');
+    document.body.removeChild(blankCanvas);
+    resizeGrip.focus();
```
Comment (thibaudcolas) [inline client/src/includes/sidePanel.js:143]: Not sure I understand why we’re doing this either. Could you add a comment?

File: client/src/includes/sidePanel.js [hunk 6]
```diff
@@ -99,6 +102,123 @@ export default function initSidePanel() {
...
+
+  const setSidePanelWidth = (width) => {
+    const { minWidth, range, width: currentWidth } = getSidePanelWidthStyles();
+    const newWidth = parseFloat(width)
+      ? Math.max(minWidth, Math.min(width, minWidth + range))
+      : currentWidth;
+
+    sidePanelWrapper.style.width = `${newWidth}px`;
+    widthInput.value = ((newWidth - minWidth) / range) * 100;
+    localStorage.setItem('wagtail:side-panel-width', newWidth);
```
Comment (thibaudcolas) [inline client/src/includes/sidePanel.js:127]: This will need to be wrapped in a `try/catch` so the code doesn’t explode when localStorage is unavailable.

File: client/src/includes/sidePanel.js [hunk 7]
```diff
@@ -99,6 +102,123 @@ export default function initSidePanel() {
...
+    // sometimes the event fires with screenX of 0, so ignore it.
+    if (!e.screenX) return;
+
+    const { width } = getSidePanelWidthStyles();
+    const newWidth = width - e.offsetX;
+    setSidePanelWidth(newWidth);
+  });
+
+  // Prevent any default dragover behaviour on the document.
+  document.addEventListener('dragover', (e) => e.preventDefault());
```
Comment (thibaudcolas) [inline client/src/includes/sidePanel.js:157]: Again this comment just describes the code rather than explain why we do this. Could you explain the "why" instead?

Is there any way we could only prevent those events if they come from the resize grip in particular? Doing this for all dragover events feels like it could have unintended consequences.

File: client/scss/components/_form-side.scss [hunk 8]
```diff
@@ -49,12 +61,38 @@
...
   }
 
+  &__resize-grip-container {
+    @apply w-absolute w-place-items-center w-hidden sm:w-flex w-z-10;
+
+    height: calc(100% - theme('spacing.8'));
+    inset-inline-start: -21px;
+  }
+
+  &__resize-grip {
```
Comment (thibaudcolas) [inline client/scss/components/_form-side.scss:71]: It would be nice if the icon within this was centered and the whole thing was a square. Right now there is more space at the bottom than top:

resize-grip

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
Working great but there are a few things I don’t understand in the code, and a few issues.

In addition there are two bits of behavior I couldn’t explain. With the keyboard, the rate of change feels too slow (could we set a higher `step`? 10 perhaps?) and isn’t constant. With a constant press on the key, I get a rapid acceleration at the very end:

keyboard

If this is just macOS key-repeat behavior, then I think setting a higher step will also solve this.

And with dragging, I often get a different cursor appearing briefly:

dragging-cursor

File: client/scss/components/_preview-panel.scss (hunk 3 above)
Comment (laymonage) [reply client/scss/components/_preview-panel.scss]: Yes and it didn't work, but we don't need this with the new implementation that doesn't use the Drag API.

File: client/scss/components/_form-side.scss (hunk 1 above)
Comment (laymonage) [reply client/scss/components/_form-side.scss:27]: This was needed so that the resize grip isn't cut off because it overflows the parent. I've moved the `w-overflow-y-auto` and paddings to `__panel` to fix the scrolling.

File: client/scss/components/_form-side.scss (hunk 2 above)
Comment (laymonage) [reply client/scss/components/_form-side.scss]: I've switched to a different approach by negating the input's value and using `aria-valuetext` so that screen readers can read the actual width value in pixels.

File: client/src/includes/sidePanel.js (hunk 4 above)
Comment (laymonage) [reply client/src/includes/sidePanel.js]: This is no longer needed in the new implementation.

File: client/src/includes/sidePanel.js (hunk 5 above)
Comment (laymonage) [reply client/src/includes/sidePanel.js]: This is no longer needed in the new implementation.

File: client/src/includes/sidePanel.js (hunk 7 above)
Comment (laymonage) [reply client/src/includes/sidePanel.js]: This is no longer needed in the new implementation.

File: client/scss/components/_form-side.scss [hunk 9]
```diff
@@ -49,12 +59,38 @@
...
   }
 
+  &__resize-grip-container {
+    @apply w-absolute w-place-items-center w-hidden md:w-flex w-z-10 w-left-[-22px];
+
+    height: calc(100% - theme('spacing.8'));
+  }
+
+  &__resize-grip {
+    @apply w-text-primary hover:w-text-primary-200 w-border w-border-transparent w-rounded w-bg-white w-p-3 w-hidden w-touch-pinch-zoom w-cursor-ew-resize;
```
Comment (thibaudcolas) [inline client/scss/components/_form-side.scss:69]: What do you think of shifting this one pixel to the right? It feels to me slightly better aligned this way, though with an even width aiming for the exact middle is hopeless.

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
Looking and working great. Tested in latest Firefox, Chrome, Safari, and iOS Safari 15.

One point I think needs addressing (`try…catch`). The other points at your discretion.

File: client/src/includes/sidePanel.js [hunk 10]
```diff
@@ -107,8 +194,9 @@ export default function initSidePanel() {
       if (!inExplorer && sidePanelOpen) {
         setPanel(sidePanelOpen);
       }
+      setSidePanelWidth(localStorage.getItem('wagtail:side-panel-width'));
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/src/includes/sidePanel.js:197]: This is still within a `try` block we had for `setPanel` above

File: client/scss/components/_form-side.scss (hunk 9 above)
Comment (laymonage) [reply client/scss/components/_form-side.scss:69]: I think I previously used `w-left-[-21px]` (so, one pixel to the right) and felt it's slightly better aligned with `-22px`. Shall we go with `-21.5px`? Not sure if half a pixel make sense, but [it seems valid](https://stackoverflow.com/questions/9080633/can-a-css-pixel-be-a-fraction) 😛

File: client/scss/components/_form-side.scss [hunk 11]
```diff
@@ -49,12 +59,38 @@
...
+
+    .form-side--open & {
+      @apply w-flex;
+    }
+
+    &:focus-within {
+      @include focus-outline;
+    }
```
[the remark this answers is not in the document]
Comment (laymonage) [reply client/scss/components/_form-side.scss]: Good call, was a bit hesitant but I realised that Safari only started supporting this in 15.4 and we support last 3 versions, so this is necessary.

File: client/scss/components/_form-side.scss (hunk 9 above)
Comment (thibaudcolas) [reply client/scss/components/_form-side.scss:69]: Let’s ask a design person what they think!

**Review by thibaudcolas — APPROVED** [review]
Thank you @laymonage! My bad re try…catch. 

I made a quick recorded demo of the functionality to try to get more into the habit of demo videos: https://www.loom.com/share/ca6baca5a85e42c3a6fec27fd21ae148
